### PR TITLE
mcp3008: add implementation for MCP3008 ADC with SPI interface

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds implementation for MCP3008 ADC with SPI interface.

### CODE REVIEW

File: mcp3008/mcp3008.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	CH3 *ADCPin
+	CH4 *ADCPin
+	CH5 *ADCPin
+	CH6 *ADCPin
+	CH7 *ADCPin
+}
+
+// ADCPin is the implementation of the ADConverter interface.
+type ADCPin struct {
+	machine.ADC
```
Comment (aykevl) [inline mcp3008/mcp3008.go]: Why do you need to embed this type? Why not just `machine.Pin`?

File: mcp3008/mcp3008.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	tx  []byte
+	rx  []byte
+	CH0 *ADCPin
+	CH1 *ADCPin
+	CH2 *ADCPin
+	CH3 *ADCPin
+	CH4 *ADCPin
+	CH5 *ADCPin
+	CH6 *ADCPin
+	CH7 *ADCPin
```
Comment (aykevl) [inline mcp3008/mcp3008.go]: Why not use an array or slice here?

Also, you could make these non-pointers, which is somewhat lighter on the heap/GC. For example:

```go
type Device struct {
    // ...
    [8]CH ADCPin
}
```

You can always get the pointer with `&device.CH[0]` for example.

File: mcp3008/mcp3008.go (hunk 1 above)
Comment (deadprogram) [reply mcp3008/mcp3008.go]: A future PR will add an interface type `ADConverter` so we can use the same as a normal `machine.ADC` for other drivers such as the `thermistor`.

File: mcp3008/mcp3008.go (hunk 2 above)
Comment (deadprogram) [reply mcp3008/mcp3008.go]: I was trying to make this act as much as possible like a "regular" `machine.ADC`, hence the specific members. The pointer is so we can in the future use as a param of type `machine.ADConverter`.

File: mcp3008/mcp3008.go (hunk 1 above)
Comment (aykevl) [reply mcp3008/mcp3008.go]: But you wouldn't need to embed this type to implement an interface?

For example, the interface could look like this:

```go
type ADCer interface {
    Get() uint16
}
```

In that case, the `ADCPin` type already implements the interface.
(Also, perhaps `Get() uint16` is a bit too generic if ADC is used as an interface).

Comment (deadprogram) [reply mcp3008/mcp3008.go]: Changed to simply `machine.Pin`

File: mcp3008/mcp3008.go (hunk 2 above)
Comment (deadprogram) [reply mcp3008/mcp3008.go]: Got rid of pointers, but left as individual members.

**Comment by deadprogram** [comment]
Changes made and force pushed.

**Pull request merged**